src/main/java/org/truffleruby/core/time/RubyDateFormatter.java

```diff
@@ -544,6 +568,168 @@ public static RopeBuilder formatToRopeBuilder(List<Token> compiledPattern, Zoned
+                        return false;
+                    }
+                    break;
+                case FORMAT_OUTPUT:
+                    RubyTimeOutputFormatter formatter = (RubyTimeOutputFormatter) token.getData();
+
+                    // Check for the attributes present in the default case
+                    if (!formatter.flags.isEmpty()) {
+                        return false;
+                    }
+                    if (formatter.width != 6) {
+                        return false;
+                    }
+
+                    // FORMAT_NANOSEC should always come after FORMAT_OUTPUT
+                    if (compiledPattern[i + 1].getFormat() != Format.FORMAT_NANOSEC) {
+                        return false;
+                    }
+                    break;
+                case FORMAT_STRING:
+                case FORMAT_DAY:
+                case FORMAT_HOUR:
+                case FORMAT_MINUTES:
+                case FORMAT_MONTH:
+                case FORMAT_SECONDS:
+                case FORMAT_YEAR_LONG:
+                case FORMAT_NANOSEC:
+                    break;
+                default:
+                    return false;
+            }
+        }
+
+        return true;
+    }
+
+    @ExplodeLoop
+    public static ManagedRope formatToRopeBuilderFast(Token[] compiledPattern, ZonedDateTime dt, Object zone,
+            RubyContext context, RubyLanguage language, Node currentNode, ErrnoErrorNode errnoErrorNode) {
+        try {
+            ManagedRope rope = null;
+
+            for (Token token : compiledPattern) {
+                final ManagedRope appendRope;
+
+                switch (token.getFormat()) {
+                    case FORMAT_ENCODING:
+                    case FORMAT_OUTPUT:
+                        continue;
+
+                    case FORMAT_STRING:
+                        appendRope = token.getRope();
+                        break;
+                    case FORMAT_DAY:
+                        appendRope = RopeConstants.paddedNumber(dt.getDayOfMonth());
+                        break;
+                    case FORMAT_HOUR:
+                        appendRope = RopeConstants.paddedNumber(dt.getHour());
+                        break;
+                    case FORMAT_MINUTES:
+                        appendRope = RopeConstants.paddedNumber(dt.getMinute());
+                        break;
+                    case FORMAT_MONTH:
+                        appendRope = RopeConstants.paddedNumber(dt.getMonthValue());
+                        break;
+                    case FORMAT_SECONDS:
+                        appendRope = RopeConstants.paddedNumber(dt.getSecond());
+                        break;
+
+                    case FORMAT_YEAR_LONG: {
+                        final int value = dt.getYear();
+
+                        if (value < 1000 || value > 9999) {
+                            CompilerDirectives.transferToInterpreterAndInvalidate();
```

Comment:
This is a potential deopt loop, we need something safer, for instance a profile + boundary call, or validate the year directly in the specialization guards with some helper method.

Reply:
Done.